### DISCUSSION TITLE
TSynth patches on the Momentum

### DISCUSSION THREAD

**Discussion by patty886** [ElectroTechnique/Momentum#12] [Suggestions]
Hi, I've had a go at copying the TSynth preset patches onto the SD card for the Momentum, and need a little help.
I've been able to copy the data of the TSynth presets onto the SD card for the Momentum, and it'll recognise the list of patches once I've modified the Patchindex file with the folder name. All the TSynth patches default to the first Momentum patch, though, and looking at the code in GitHub ("SD" in the Momentum repository and "PresetPatches" in the TSynth repository) there's a different naming convention for the parameters, the number of parameters listed in the code don't match up 1:1, and the parameters aren't named in the TSynth code. If anyone knows what the parameters are and can share that'd be most appreciated, otherwise I'll keep trying and see where I get to :)

**Comment by patty886** [comment]
I have the patches translated so that they're readable in Momentum and make different sounds.
There's a lot of digital noise/artefact in quite a few of them, though, so I may not have mapped all the parameters properly. If anyone would like to take a look I can post/send the files :)

**Reply by Electrotechnique** [reply]
Yes, you can post them here as a zip. Thanks.

**Comment by patty886** [comment]
That's great, thanks very much. There's been a bit of a delay between uploading, I tried adjusting some settings (e.g. I realised that pitch LFO rate has a maximum value of 40, so adjusted the formula to account for that). There's an improvement, with less noise/artefact, but only a few patches are noise-free so I'm still missing something.
I've included the excel file where I've copied the values from the Constants.cpp file (Momentum Constants tab) to use as a lookup table for TSynth values (in the Sheet1 tab).
I couldn't match up values from the TSynth patch string for lfoTempValue, lfoSyncFreq, or midiclkTime Interval with the Momentum patch string.

Thanks very much for the help :)

[TSynth_Momentum patches.zip](https://github.com/ElectroTechnique/Momentum/files/14024526/TSynth_Momentum.patches.zip)

**Comment by Electrotechnique** [comment]
Looking the patches on Momentum, many of the parameters are wrong. The sample and hold waveform is on patches that don't require it, causing noise. Also detune and oscillator levels are wrong sometimes.  I haven't been through everything, but they all need rechecking to see if the mapping from one to the other is correct. 

On Momentum there are PitchLFOMidiClkSync, FilterLFOMidiClkSync which are yes or no values.  midiclkTime  and lfoSyncFreq are worked out from the lfo rates. and shouldn't have been stored on TSynth.

```
 case CCoscLfoRate:
    {
        currentPatch.PitchLFORate = value;
        float rate = 0.0;
        if (groupvec[activeGroupIndex]->getPitchLfoMidiClockSync())
        {
            // TODO: Multitimbral - MIDI Tempo stuff remains global?
            lfoPitchTempoValue = LFOTEMPO[value];
            rate = lfoSyncFreq * LFOTEMPO[value];
            if (!setEncValue(CCoscLfoRate, value, LFOTEMPOSTR[value]))
                showCurrentParameterOverlay(ParameterStrMap[CCoscLfoRate], LFOTEMPOSTR[value]);
        }
        else
        {
            rate = LFOMAXRATE * POWER[value];
            if (!setEncValue(CCoscLfoRate, value, String(rate) + F(" Hz")))
                showCurrentParameterOverlay(ParameterStrMap[CCoscLfoRate], String(rate) + F(" Hz"));
        }
        groupvec[activeGroupIndex]->setPitchLfoRate(rate);

        break;
    }
```

**Comment by patty886** [comment]
Thanks very much. The incorrect translation might have happened because I was looking up the TSynth constants in the file I created from the code for the Momentum constants. They looked pretty similar at first glance, but I'll redo it using the TSynth constants file and see what I get.
I took the waveform numbers straight across from the TSynth values because the synth_waveform.cpp and synth_waveform.h files are identical between TSynth and Momentum. The values for the waveform (A or B) are double-digit in the Momentum, which I'm presuming are on a scale from 0-127 similar to values for other parameters. I can't find a lookup to translate the TSynth patch values for waveform into the values for Momentum, though.
Am I off the mark trying to translate a lot of these by turning the Constants.cpp file into an excel table and running vlookups to get the MIDI (0-127) value? It looked like the values in the Momentum patches were entered as a number between 0-127 (0-255 for a couple of parameters, e.g. filter frequency), whereas for TSynth some values were whole numbers and others were decimals on a scale of 0 to 1.

**Reply by Electrotechnique** [reply]
Yes, much of the values are based on 0-127 to map to MIDI messages. Filter cutoff is 0-255 (but maps to 0-127 MIDI). Some values like pitchbend range are the actual (absolute) value. TSynth uses absolute values for parameters because it has potentiometers that will give absolute values for their position, which are easier to map to parameter values. Momentum uses encoders that increment and decrement a relative value, and so it was decided that the patch files would be better (easier) with relative values. The disadvantage is that changes such as adding new waveforms, will mean workarounds for existing patchfiles.

**Comment by patty886** [comment]
> Looking the patches on Momentum, many of the parameters are wrong. The sample and hold waveform is on patches that don't require it, causing noise.
> 

Thanks for all your help so far, I've translated the waveforms as best I can, but I'm having trouble matching up some of them.

I haven't found any waveforms on Momentum that correlate directly with the bandlimit waveforms (bandlimit sawtooth, bandlimit sawtooth reverse, bandlimit square, and bandlimit pulse), am I right in thinking that variable triangle should sub for bandlimit sawtooth and bandlimit sawtooth reverse, and variable pulse should sub for bandlimit square and bandlimit pulse?

There are also a different number of icons on the TSynth front panel compared to the number of options in the code (8 options on the front panel, 11 options in the code for Osc1 and 9 options in the code for Osc2). 
The TSynth front panel doesn't have options to select a sine or arbitrary waveform, but it's in the code for a number of patches. None of the Momentum patches show sine or arbitrary as an option when I call them up so I can't translate this into values in the Momentum code directly
I thought at first that the sine and arbitrary waveforms in the synth_waveforms.h file correspond to the parabolic and harmonic waves, respectively. In the TSynth VoiceGroup.h file the values 103 and 104 correspond to parabolic and harmonic in lines 28-29, but then reference back to the arbitrary waveform in lines 219-235 and 240-256.

Is there some code I'm missing that sets out which range of MIDI values in Momentum translate to which waveforms? The synth_waveforms.h file is the only one I can find but it says it's based on TSynth 2.3.1
I think I've found something in lines 253 to 386 of the TSynth.ino code that might correlate, I'll investigate that.

> Also detune and oscillator levels are wrong sometimes. I haven't been through everything, but they all need rechecking to see if the mapping from one to the other is correct.
> 

These seemed to live in a range from 0 to 1 in the TSynth patches, so I assumed that correlated to 0 to 127 and multiplied the value in the TSynth code by 127 and rounded it to the nearest whole number. Is there another way I should tackle it? I couldn't find anything in the Detune.cpp or Detune.h to translate simply. The Detune.cpp file has arrays of values, but not all of the values are present in the array (0.98 and 0.99 are the most common values in the TSynth patches, but these don't appear in any of the arrays in the Detune.h file

**Reply by patty886** [reply]
Much better luck this time, I've used the value ranges for the oscillator waveforms and LFO waveforms in lines 253 to 386 of the TSynth.ino code and changed the Momentum patch values for Waveform A, Waveform B, Pitch LFO Waveform, and LFO Waveform (as the FilterLFOWaveform equivalent).

I'm still not quite sure what to do about the patches that have a TSynth value of "0"; that's meant to correlate to a sine wave, but I can't find ranges in the TSynth.ino code for sine wave in the oscillator waveforms. There's a range for filter waveform, but not oscillator that I can find.
I've left it as "0" for the patches that have it, which switches off that oscillator, I can play around with it in the patches from here.

I think some of the detune settings and filter and amp ADSR settings aren't spot on, at least for my taste, but I can adjust those in the patches and save once it's adjusted to taste.
I did find that the octave range of the keyboard makes a big difference when auditioning patches, which is good to know!

Revised patch settings attached if you'd like to have a look :)

[TSynth_Momentum patches_2.zip](https://github.com/ElectroTechnique/Momentum/files/14890562/TSynth_Momentum.patches_2.zip)
